### ISSUE TITLE
Parleyview.tableview crash

### ISSUE THREAD

**Issue by rubenaalders** [parley-messaging/ios-library#28]
I'd like to confirm that issue #26 is fixed. Unfortunately, this introduced another crash I think. I attached the relevant stacktrace below.

It feels like this happens at random. Sometimes immediately, sometimes after a few seconds. If you need more information please ask.

> 0   wehkampDependencies           	0x0000000106db0c98 0x106d04000 + 707736
1   wehkampDependencies           	0x0000000106db03e8 0x106d04000 + 705512 (ParleyView.swift:430)
2   wehkampDependencies           	0x0000000106db0d68 @objc ParleyView.tableView(_:cellForRowAt:) + 140 (<compiler-generated>:0)
3   wehkampDependencies           	0x0000000106db5704 0x106d04000 + 726788 (ReversedTableView.swift:103)
4   wehkampDependencies           	0x0000000106db583c @objc ReversedTableView.tableView(_:cellForRowAt:) + 140 (<compiler-generated>:0)
5   UIKitCore                     	0x00000001c4091284 -[UITableView _createPreparedCellForGlobalRow:withIndexPath:willDisplay:] + 1256 (UITableView.m:16516)
6   UIKitCore                     	0x00000001c42223b8 -[UITableView _updateVisibleCellsForRanges:createIfNecessary:] + 596 (UITableView.m:3111)
7   UIKitCore                     	0x00000001c40af7d8 -[UITableView _updateVisibleCellsNow:] + 1088 (UITableView.m:3253)
8   UIKitCore                     	0x00000001c3fb8d28 -[UITableView layoutSubviews] + 148 (UITableView.m:10575)
9   UIKitCore                     	0x00000001c3fc6ab0 -[UIView(CALayerDelegate) layoutSublayersOfLayer:] + 1980 (UIView.m:18473)
10  QuartzCore                    	0x00000001c5683fec CA::Layer::layout_if_needed(CA::Transaction*) + 500 (CALayer.mm:10204)
11  QuartzCore                    	0x00000001c5675e58 CA::Layer::layout_and_display_if_needed(CA::Transaction*) + 148 (CALayer.mm:2497)
12  QuartzCore                    	0x00000001c568a4a0 CA::Context::commit_transaction(CA::Transaction*, double, double*) + 456 (CAContextInternal.mm:2689)
13  QuartzCore                    	0x00000001c5693628 CA::Transaction::commit() + 652 (CATransactionInternal.mm:431)
14  QuartzCore                    	0x00000001c5675538 CA::Transaction::flush_as_runloop_observer(bool) + 88 (CATransactionInternal.mm:939)
15  UIKitCore                     	0x00000001c42b920c _UIApplicationFlushCATransaction + 52 (UIApplication.m:3262)
16  UIKitCore                     	0x00000001c44dd6d0 _UIUpdateSequenceRun + 84 (_UIUpdateSequence.mm:112)
17  UIKitCore                     	0x00000001c4b07fdc schedulerStepScheduledMainSection + 172 (_UIUpdateScheduler.m:916)
18  UIKitCore                     	0x00000001c4b07624 runloopSourceCallback + 92 (_UIUpdateScheduler.m:1008)
19  CoreFoundation                	0x00000001bd0f1e40 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 28 (CFRunLoop.c:1974)
20  CoreFoundation                	0x00000001bd10297c __CFRunLoopDoSource0 + 176 (CFRunLoop.c:2018)
21  CoreFoundation                	0x00000001bd03e72c __CFRunLoopDoSources0 + 244 (CFRunLoop.c:2055)
22  CoreFoundation                	0x00000001bd044020 __CFRunLoopRun + 836 (CFRunLoop.c:2970)
23  CoreFoundation                	0x00000001bd057bac CFRunLoopRunSpecific + 612 (CFRunLoop.c:3390)
24  GraphicsServices              	0x00000001df94b360 GSEventRunModal + 164 (GSEvent.c:2196)
25  UIKitCore                     	0x00000001c4297a34 -[UIApplication _run] + 888 (UIApplication.m:3730)
26  UIKitCore                     	0x00000001c409de04 UIApplicationMain + 340 (UIApplication.m:5320)
27  wehkamp                       	0x0000000102d0c9b0 main + 68 (AppDelegate.swift:21)
28  dyld                          	0x000000024f9018f8 start + 2296 (dyldMain.cpp:1145)

**Comment by alexkok** [comment]
Hi @rubenaalders, seems like this is a different issue. We cannot seem to reproduce this one yet, can you tell us a bit more about your setup?

A few questions:
- Do you have **Offline Messaging** enabled with a (custom) **ParleyDataSource**? If so, can you try without it?
- Does this also happen for you inside the Parley Demo project that is inside this repo?
- Does Parley inside the app use the **Polling** mechanism or **Push Messages** mechanism when chatting?
- Which device / iOS version does this happen on? (I assume physical device)

**Comment by rubenaalders** [comment]
Due to priorities this took me a bit longer to get too. This is what I observed:

1. We do not. All we do is put the userdata in `Parley.setUserInformation()` and style the chat page
2. I don't really have time to test it right now, but will check soon
3. We use the push message mechanism
4. The test was done on an iPhone 13 mini with ios 15. Although on a simulator I get other issues (array out of bounds on ParleyView line 430) with loading messages because of the push message mechanism not working on simulators.

**Comment by alexkok** [comment]
Thanks, we'll continue to investigate (haven't reproduced it in our demo as of yet though).

2 additional questions:
- What kind of messages are inside the chat when this happens? (just text ones, or do they also already include carousel/quickreplies/etc.)? And (approximately) how many messages are inside it now?
- Does this happen with new, empty chats as well? (without `setUserInformation()` for example)

**Comment by alexkok** [comment]
Hi @rubenaalders, we've investigated the behaviour and resolved some issues and crashes we've encountered. Under which the crash in this issue was seen as well.

Sidenote: Make sure to call `Parley.configure()` only once during the lifetime of the app (calling it multiple times is unexpected and not needed)

Question: Do you call `Parley.setUserInformation()` before or after calling `configure()`?

Please check with the version that we just released: 3.6.0 🚀  and let us know whether this resolves the crashes you were facing.

**Comment by rubenaalders** [comment]
Hi @alexkok,

Sorry for the late reply.

We unfortunately set it up so that the Parley.configure() is called everytime the user gets to the chat screen as that is it's own part of the app. 3.6.0 seems to have indeed fixed all the crashes. We now only have issues with overlapping buttons on the text as can be seen in the screenshot. Any idea what could be the culprit?

![Simulator Screen Shot - iPhone 12 - 2022-08-10 at 09 01 56](https://user-images.githubusercontent.com/4243098/183836618-558ddc85-b5f0-4080-9b57-a653c570b528.png)

**Comment by alexkok** [comment]
Hi @rubenaalders 

Thanks for your response! Sorry for the delay, I was on holiday. 

That partly explains the crashes we saw happening

Please note that Parley is designed to only call the `configure` method only once over the lifetime of the app (as stated in the readme). Calling this more times could result in unexpected behavior, although it works. The configure method is there such that Parley can be initialized before showing the chat, providing the best user experience. While when calling configure each time when opening the chat, the user will likely see the loading state as well.

We found an issue causing the quick replies not to render as expected in every case, under which when opening a chat that has quick replies.

Should be fixed in version 3.6.1 🚀 

I'll close this issue (this was more a separate one as well) -- Feel free to create a new issue if you find anything 👍
